src/ripple/app/misc/impl/ValidatorList.cpp

```diff
@@ -202,80 +205,90 @@ ValidatorList::applyList (
+                    continue;
+                }
+
+                std::string domain;
+
+                if (val.isMember(jss::domain) && val[jss::domain].isString())
+                    domain = val[jss::domain].asString();
+
+                publisherList.emplace_back(
+                    PublicKey{Slice{ret.first.data(), ret.first.size()}},
+                    domain);
             }
 
-            if (val.isMember ("manifest") && val["manifest"].isString ())
-                manifests.push_back(val["manifest"].asString ());
+            if (val.isMember("manifest") && val["manifest"].isString())
+                manifests.push_back(val["manifest"].asString());
         }
     }
 
-    // Update keyListings_ for added and removed keys
-    std::sort (
-        publisherList.begin (),
-        publisherList.end ());
-
-    auto iNew = publisherList.begin ();
-    auto iOld = oldList.begin ();
-    while (iNew != publisherList.end () ||
-        iOld != oldList.end ())
     {
-        if (iOld == oldList.end () ||
-            (iNew != publisherList.end () &&
-            *iNew < *iOld))
+        auto comparator = [](ValidatorEntry const& a, ValidatorEntry const& b)
         {
-            // Increment list count for added keys
-            ++keyListings_[*iNew];
-            ++iNew;
+            return a.pubkey < b.pubkey;
+        };
+
+        // Update keyListings_ for added and removed keys
+        std::sort(publisherList.begin(), publisherList.end(), comparator);
+
+        {
+            // Removed elements appear in the old list but not in the new
+            std::vector<ValidatorEntry> removed;
+            std::set_difference(oldList.begin(), oldList.end(),
+                publisherList.begin(), publisherList.end(),
+                std::back_inserter(removed), comparator);
+            for (auto const& r : removed)
+                keyListings_[r.pubkey]--;
         }
-        else if (iNew == publisherList.end () ||
-            (iOld != oldList.end () && *iOld < *iNew))
+
         {
-            // Decrement list count for removed keys
-            if (keyListings_[*iOld] <= 1)
-                keyListings_.erase (*iOld);
-            else
-                --keyListings_[*iOld];
-            ++iOld;
+            // Added elements appear in the new list but not in the old
+            std::vector<ValidatorEntry> added;
+            std::set_difference(publisherList.begin(), publisherList.end(),
+                oldList.begin(), oldList.end(), std::back_inserter(added),
+                comparator);
+            for (auto const& a : added)
+                keyListings_[a.pubkey]++;
         }
-        else
+
+        // Remove all entries that are not longer referenced
+        for (auto it = keyListings_.begin(); it != keyListings_.end();)
         {
-            ++iNew;
-            ++iOld;
+            if (it->second <= 0)
+                it = keyListings_.erase(it);
+            else
+                ++it;
```

Comment:
While the new code is more readable, and I like the use of the std::algorithms, this turns a 1-pass algorithm into a 3-pass algorithm.  I'm not sure it is worth the performance hit.

Reply:
I share your concern generally, but I don't think this code is invoked frequently enough to really matter. And the new code is, as you say, much more readable and (imo) less error prone. I had to read the existing code a couple of times to _really_ convince myself it does what it claims to.